### ISSUE TITLE
Interaction with bubbletea

### ISSUE THREAD

**Issue by flowchartsman** [bitfield/script#178]
Perhaps similar in spirit to some aspects of #116, I think you could provide a set of simple, interactive inputs using TUI elements provided by [bubbletea](https://github.com/charmbracelet/bubbletea) and [bubbles](https://github.com/charmbracelet/bubbles).  There are a couple of different approaches that you could take, or combine. A compelling one is a technique I saw used in [chezmoi](https://github.com/twpayne/chezmoi) where templates receive a funcmap that provides functions which, when executed, create a bubbles widget on the fly and interpolate its value.  Here's a quick (and dirty) prototype.

```go
package main

import (
	"bytes"
	"fmt"
	"log"
	"os"
	"text/template"

	"github.com/charmbracelet/bubbles/textinput"
	tea "github.com/charmbracelet/bubbletea"
)

func main() {
	tmpl, err := template.New("").Funcs(interactiveFns()).Parse(os.Args[1])
	if err != nil {
		log.Fatalf("parsing tempalte: %v", err)
	}
	var buf bytes.Buffer
	tmpl.Execute(&buf, nil)
	os.Stdout.Write(buf.Bytes())
}

type interactiveCtx struct {
	values map[string]any
}

func interactiveFns() template.FuncMap {
	c := &interactiveCtx{
		values: map[string]any{},
	}
	return template.FuncMap{
		"stringInput": stringInput(c),
	}
}

func stringInput(c *interactiveCtx) func(string) (any, error) {
	return func(inputPrompt string) (any, error) {
		if val, found := c.values[inputPrompt]; found {
			return val, nil
		}
		p := tea.NewProgram(initialModel(inputPrompt))
		finalModel, err := p.Run()
		if err != nil {
			return nil, err
		}
		value := finalModel.(textInputModel).Value()
		c.values[inputPrompt] = value
		return value, nil
	}
}

type (
	errMsg error
)

type textInputModel struct {
	prompt    string
	textInput textinput.Model
	err       error
}

func initialModel(inputPrompt string) textInputModel {
	ti := textinput.New()
	ti.Focus()
	ti.CharLimit = 156
	ti.Width = 20

	return textInputModel{
		prompt:    inputPrompt,
		textInput: ti,
		err:       nil,
	}
}

func (m textInputModel) Init() tea.Cmd {
	return textinput.Blink
}

func (m textInputModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	var cmd tea.Cmd

	switch msg := msg.(type) {
	case tea.KeyMsg:
		switch msg.Type {
		case tea.KeyEnter, tea.KeyCtrlC, tea.KeyEsc:
			return m, tea.Quit
		}

	case errMsg:
		m.err = msg
		return m, nil
	}

	m.textInput, cmd = m.textInput.Update(msg)
	return m, cmd
}

func (m textInputModel) View() string {
	return fmt.Sprintf(
		"%s %s",
		m.prompt,
		m.textInput.View(),
	) + "\n"
}

func (m textInputModel) Value() string {
	return m.textInput.Value()
}

```

Example input/output
```
go run main.go 'User {{stringInput "First Name"}} {{stringInput "Last Name"}} has a first name of {{
stringInput "First Name"}}'
First Name > Rob                  
Last Name > Robertson            
User Rob Robertson has a first name of Rob
```

**Comment by bitfield** [comment]
Thanks for the suggestion, @flowchartsman! It sounds interesting. How, specifically, do you think the `script` package should provide support for this idea? Maybe you could give an example of a `script`-based program that shows what you have in mind.

**Comment by flowchartsman** [comment]
Yet another option would be a two-stage approach with multiple delimiter types, if you wanted to keep user inputs visually separate like:

```go
script.Exec("pulsar-admin sources create --name ${source name} --destination-topic ${source topic} --type ${source sourcetype}").WriteFile("create_"+newSource["name"]+".log)
```
```

**Comment by flowchartsman** [comment]
> I'm not sure if there's an obvious way to integrate it into the model of "everything is a method on a pipe". What do you think?

Yes, I think the template angle fits a bit more naturally for the "anywhere in the pipe" situation, though it's less in keeping with the functional-style shell pipe analogy, since it introduces side-effects.  I'm am trying to think of any good examples from unixland where steps in a pipeline break out to ask for input, but I'm not actually coming up with any, which might be a good argument to have it apart from the pipeline.

Having said that, if you think of this as a library that wants to provide a clean API to do things users might otherwise want to do with shell scripts, then interactivity is still useful  to provide, since it's something a lot of users want to have in scripts, but avoid because the boilerplate to do it in a script is kind of a pain.

When I was filing this issue, I was actually inpspired in part by the [gum](https://github.com/charmbracelet/gum) tool, which tries to bridge that gap, and lets script writers do things like `COMMIT_TYPE=$(gum choose "fix" "feat" "docs" "style" "refactor" "test" "chore" "revert")`. I can see this being super useful for personal tooling, especially with flag overriding, and if the design is clean, it could really expand the types of programs that could be written with the`script`, without changing the core workflow.

It could certainly work as an add-on, but for my part I'd be more inclined to just wrap the package into one of my own to give me access to all of the tools at once, so that I could have a go-to dependency for most of the tasks of this type that I want to write.

**Comment by flowchartsman** [comment]
the other inspiration is that I wrote a tool recently to fill a similar niche to `shell` to let me easily automate repetitive tasks, and I found myself wanting an way to do inputs too, so that I could just reach for that one library any time I needed. I also made the decision to have it be YAML-driven, but I didn't like the requirement of an extra file, when what I really wanted was a simple way to chain things together.

Then something in my memory clicked, and I remembered that `shell` exists, so I thought I might propose an extension here instead and get the best of both worlds :)

**Comment by bitfield** [comment]
It seems clear that there's a fairly common general requirement for 'getting user input as part of a pipeline', even if that's as simple as 'press any key to continue', or as complex as 'select from the following interactive menu'. 

I'd like to invite some specific design proposals for this, from whoever's interested or has ideas. The best way to do this is probably to comment with a `script` program you'd _like_ to write, involving user input, showing how your proposed API would work in a realistic application.

**Comment by flowchartsman** [comment]
So I've been doing some Pulsar work recently, and one of the things I wanted to do was to select a list of pulsar connectors to download locally, from the archive site, first by selecting from the available versions, then by selecting from the available connector files. With a [goquery](https://github.com/PuerkitoBio/goquery) PR, which I've been kicking around, that might look like:


```go
const (
	archiveURL = `https://archive.apache.org/dist/pulsar/`
        connectorPfx = `https://archive.apache.org/dist/pulsar/pulsar-{{.}}/connectors/`
)
func main() {
	pVersions := regexp.MustCompile(`pulsar-(?P<pversion>\d+\.\d+.\d+)/`)
	script.Get(archiveURL).GoQuery(`a`).MatchRegexp(pVersions).ChooseOneFrom("{{.match.pversion}}").
                Get(connectorPfx).GoQuery(`a`).ChooseAnyFrom("{{.}}").ExecForEach("curl -sLO {{.}}").Stdout()
}
```

**Comment by bitfield** [comment]
Sounds interesting! Perhaps we could prototype this with `gum`?

**Comment by flowchartsman** [comment]
I'm absolutely down to give it a shot. Will probably need a couple weeks before some time frees up for OS work, but I'm happy to prototype an implementation!
